crates/task-impls/src/quorum_proposal_recv/handlers.rs

```diff
@@ -72,62 +69,38 @@ async fn validate_proposal_liveness<TYPES: NodeType, I: NodeImplementation<TYPES
         warn!("Couldn't store undecided state.  Error: {:?}", e);
     }
 
-    let liveness_check = justify_qc.view_number() > consensus_write.locked_view();
-
-    let high_qc = consensus_write.high_qc().clone();
-    drop(consensus_write);
-
     // Broadcast that we've updated our consensus state so that other tasks know it's safe to grab.
     broadcast_event(
         HotShotEvent::ValidatedStateUpdated(view_number, view).into(),
         event_sender,
     )
     .await;
+
     broadcast_event(
         HotShotEvent::NewUndecidedView(leaf.clone()).into(),
         event_sender,
     )
     .await;
-
-    if liveness_check {
-        let new_view = proposal.data.view_number + 1;
-
-        // This is for the case where we form a QC but have not yet seen the previous proposal ourselves
-        let should_propose = task_state.quorum_membership.leader(new_view) == task_state.public_key
-            && high_qc.view_number() == proposal.data.view_number();
-
-        if should_propose {
-            debug!(
-                "Attempting to publish proposal after voting for liveness; now in view: {}",
-                *new_view
-            );
-            broadcast_event(
-                HotShotEvent::QuorumProposalLivenessValidated(proposal.data.clone()).into(),
-                event_sender,
-            )
-            .await;
-        }
-
-        return Some(proposal.data.clone());
-    }
-
-    None
 }
 
 /// Handles the `QuorumProposalRecv` event by first validating the cert itself for the view, and then
-/// evaluating if a liveness check is needed for the proposal, which runs when the proposal cannot be
-/// found in the internal state map.
+/// updating the states, which runs when the proposal cannot be found in the internal state map.
 ///
 /// This code can fail when:
 /// - The justify qc is invalid.
 /// - The task is internally inconsistent.
 /// - The sequencer storage update fails.
+///
+/// # Returns
+/// * `Ok(true)` if the validation succeeds.
+/// * `Ok(false)` if the validation isn't completed due to missing information.
+/// * `Err(..)` if the validation fails.
 pub(crate) async fn handle_quorum_proposal_recv<TYPES: NodeType, I: NodeImplementation<TYPES>>(
     proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
     sender: &TYPES::SignatureKey,
     event_sender: &Sender<Arc<HotShotEvent<TYPES>>>,
     task_state: &mut QuorumProposalRecvTaskState<TYPES, I>,
-) -> Result<Option<QuorumProposal<TYPES>>> {
+) -> Result<bool> {
```

Comment:
I think I preferred the other mechanism of returning the proposal. Using a boolean type doesn't convey meaning to the reader in the match statement in `mod.rs` and now I need to go to the source of the code, whereas returning the proposal conveys meaning and I immediately know that if I receive the proposal, I have to use it.

If others disagree, I'm willing to yield on this point, but I think we should at minimum define a descriptive constant such as `LIVENESS_CHECK_SUCCEEDED` or something to that effect.

Reply:
I don't have a strong opinion on this. But if we use `Result<bool>`, might need more comment at the place where this function is called (e.g. 'crates/task-impls/src/quorum_proposal_recv/mod.rs').

Reply:
I think that circumstance is precisely what I'd like to avoid. People, by choice or accident, may not comment a block of code. By making the code itself descriptive of what it's doing, it's a lot more clear what is going on.

I think this situation might highlight the necessity for the liveness check on the proposal to perhaps be its own function, handled outside of the quorum proposal recv handler entirely. It can be at the top-level by the match statement, that way we can simply return `Result<()>`, where not being able to find the parent is an error. Anyhow, unfortunately, makes specializing the error not as cut and dry as the baked-in error type, but I think I'd greatly prefer that since, ordinarily, we'd be logging an error if such were to happen (and the liveness check fails).

Reply:
Thanks for @jparr721 and @dailinsubjam's comments! Thinking about this more, I'm leaning towards replacing `Result<bool>` with an enum of three variants (similar to the `LIVENESS_CHECK_SUCCEEDED` constant suggested above) so that:
* It's more explicit to the caller that the three variants need to be handled.
* We don't have to duplicate the proposal in the return type, but it's still clear whether the check succeeds.